depthcharge/data/spectrum_datasets.py

```diff
@@ -1,711 +1,475 @@
-            grp = self._handle[str(idx)]
-            offsets.append(grp.attrs["n_spectra"])
-            self._file_map[grp.attrs["path"]] = idx
-
-        self._file_offsets = np.cumsum([0] + offsets)
-
-        # Build a map of 1D indices to 2D locations:
-        grp_idx = 0
-        for lin_idx in range(offsets[-1]):
-            grp_idx += lin_idx >= offsets[grp_idx + 1]
-            row_idx = lin_idx - offsets[grp_idx]
-            self._locs[lin_idx] = (grp_idx, row_idx)
-
-        self._offsets = None
-        _ = self.offsets  # Reinitialize the offsets.
-
-    def _validate_index(self) -> None:
-        """Validate that the index is appropriate for this dataset."""
-        preproc_hash = _hash_obj(tuple(self.preprocessing_fn))
-        with self:
-            try:
-                assert self._handle.attrs["ms_level"] == self.ms_level
-                assert self._handle.attrs["preprocessing_fn"] == preproc_hash
-                if self._annotated:
-                    assert self._handle.attrs["annotated"]
-            except (KeyError, AssertionError):
-                raise ValueError(
-                    f"'{self.path}' already exists, but was created with "
-                    "incompatible parameters. Use 'overwrite=True' to "
-                    "overwrite it."
-                )
-
-            self._reindex()
-
-    def _get_parser(
-        self,
-        ms_data_file: PathLike,
-    ) -> MzmlParser | MzxmlParser | MgfParser:
-        """Get the parser for the MS data file.
-
-        Parameters
-        ----------
-        ms_data_file : PathLike
-            The mass spectrometry data file to be parsed.
-
-        Returns
-        -------
-        MzmlParser, MzxmlParser, or MgfParser
-            The appropriate parser for the file.
-        """
-        kw_args = {
-            "ms_level": self.ms_level,
-            "valid_charge": self.valid_charge,
-            "preprocessing_fn": self.preprocessing_fn,
-        }
+            path = Path(self._tmpdir.name) / f"{uuid.uuid4()}.lance"
 
-        if ms_data_file.suffix.lower() == ".mzml":
-            return MzmlParser(ms_data_file, **kw_args)
+        self._path = Path(path)
+        if self._path.suffix != "lance":
+            self._path = path.with_suffix(".lance")
 
-        if ms_data_file.suffix.lower() == ".mzxml":
-            return MzxmlParser(ms_data_file, **kw_args)
+        # Now parse spectra.
+        if spectra is not None:
+            spectra = utils.listify(spectra)
+            batch = next(_get_records(spectra, **kwargs))
+            lance.write_dataset(
+                _get_records(spectra, **kwargs),
+                self._path,
+                mode="overwrite",
+                schema=batch.schema,
+            )
 
-        if ms_data_file.suffix.lower() == ".mgf":
-            return MgfParser(ms_data_file, **kw_args)
+        elif not self._path.exists():
+            raise ValueError("No spectra were provided")
```

Comment:
why isnt this a "FileNotFoundError" ??

Reply:
Well this one is a little weird. The `from_lance()` method requires that the user provide an existing lance database, but if you want to init a new dataset, you don't need to provide the path to one. So we have two real use cases for `__init__()`:

1. A fresh initialization - the user must provide spectra, but no the lance file.
2. Loading a previous lance dataset - calls init without any spectra, but requires a lance file. 

This error actually assumes that it arose during case 1, where the user did not provide spectra to insert into the lance dataset.